docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -0,0 +1,198 @@
+
+### Snapshot Chunk Format
+
+The Gaia data structure consists of a set of named IAVL trees. A root hash is constructed by taking the root hashes of each of the IAVL trees, then constructing a Merkle tree of the sorted name/hash map.
+
+IAVL trees are versioned, but a snapshot only contains the version relevant for the snapshot height. All historical versions are ignored.
+
+IAVL trees are insertion-order dependent, so key/value pairs must be stored in an appropriate insertion order to produce the same tree branching structure and thus the same Merkle hashes.
+
+A chunk corresponds to a subtree key range within an IAVL tree, in insertion order, along with the Merkle proof from the root of the subtree all the way up to the root of the multistore (including the Merkle proof for the surrounding multistore.
+
+```go
+struct SnapshotChunk {
+    Store   string            // Name (key) of IAVL store in outer MultiStore
+    Keys    [][]byte          // Snapshotted keys in insertion order
+    Values  [][]byte          // Snapshotted values corresponding to Keys
+    Proof   []merkle.ProofOp  // Merkle proof from subtree root to MultiStore root
+}
+```
+
+This chunk structure is believed to be sufficient to reconstruct an identical IAVL tree by applying separate chunks, but this may require the chunks to be ordered in a certain way; further research is needed.
+
+We do not use IAVL RangeProofs, since these include redundant data such as proofs for intermediate and leaf nodes that can be derived from the above data.
+
+Chunks should be built greedily by collecting key/value pairs constituting a complete subtree up to some size limit (e.g. 16 MB), then serialized and stored in the file system as `snapshots/<height>/<format>/<chunk>/data` and served via `RequestGetSnapshotChunk`.
+
+### Snapshot Scheduling
+
+Snapshots should be taken at some configurable height interval, e.g. every 1000 blocks. All nodes should preferably have the same snapshot schedule, such that all nodes can serve chunks for a given snapshot.
+
+Taking consistent snapshots of IAVL trees is greatly simplified by them being versioned: simply snapshot the version that corresponds to the snapshot height, while concurrent writes create new versions. IAVL pruning must not prune a version that is being snapshotted.
+
+Snapshots must also be garbage collected after some configurable time, e.g. by keeping the latest `n` snapshots.
+
+## Open Questions
+
+* Is it possible to reconstruct an identical IAVL tree given separate subtrees in an appropriate order, or is more data needed about the branch structure?
+
+* Should we punish nodes that provide invalid snapshots?
+
+* Should `ResponseOfferSnapshot` and `ResponseApplySnapshotChunk` return specific codes for e.g. invalid heights, unknown formats, and so on?
```

Comment:
probably not if we assume logging (ability to figure out why certain things are not happening) on the app side

Reply:
Sure, but Tendermint can do different things in response. `ResponseOfferSnapshot` returns an internal error, it may make sense to either abort or retry, while a verification error should probably indicate that the sender can't be trusted. I'll be adding status codes for these now.

Reply:
> If ResponseOfferSnapshot returns an internal error, it may make sense to either abort or retry, while a verification error should probably indicate that the sender can't be trusted

that makes sense